feat: Split NFA positive tags into start and end transitions to encapsulate a capture group.

SharafMohamed opened the pull request:
# Description
- In tagged Regex, positive tags can exist in isolation. However, for capture groups to be useful we always need them to exist in pairs, when before the capture group and one after.
- Split the existing positive tags into a start and end tag.

# Validation performed
- Unit-tests updated to output both start and end tags, runs correctly.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Improved handling of capture groups in regex processing, enhancing NFA state transitions.
	- Clearer distinction between start and end tagged transitions for better clarity in transition management.

- **Bug Fixes**
	- Enhanced traversal logic for tagged transitions to ensure accurate processing.

- **Tests**
	- Updated test cases to reflect changes in serialized NFA output, including clearer naming conventions for transitions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SharafMohamed commented on src/log_surgeon/finite_automata/TaggedTransition.hpp:
```diff
@@ -64,7 +64,7 @@ class NegativeTaggedTransition {
      * @throw std::invalid_argument if any elements in `tags` is `nullptr`.
      */
-    NegativeTaggedTransition(std::vector<Tag const*> tags, NFAStateType const* dest_state)
+    NegativeTaggedTransition(std::vector<Tag*> tags, NFAStateType* dest_state)
```
Changed it back to const.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -894,9 +894,16 @@ template <typename NFAStateType>
             = nfa->new_state_with_positive_tagged_transition(m_tag.get(), end_state);
+    nfa->set_root(capture_group_start_state);
     m_group_regex_ast->add_to_nfa_with_negative_tags(nfa, state_with_positive_tagged_transition);
+
```
```suggestion
```

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -894,9 +894,16 @@ template <typename NFAStateType>
 template <typename NFAStateType>
 void RegexASTCapture<NFAStateType>::add_to_nfa(RegexNFA<NFAStateType>* nfa, NFAStateType* end_state)
```
iiuc, this should be the schematics of the constructed capture group nfa:
```
root --(pos_tagged_start_transition)--> capture_group_start_state --> [inner capture group NFA] --(neg_tagged_transition)--> neg_state --> capture_group_end_state --(pos_tagged_end_transition)--> end (accepting)
```
Please correct me if it's wrong. Also we could add a comment string to document this diagram

SharafMohamed replied:
Pretty much. End is not necessarily accepting. It would look like this: 
`root --(pos_tagged_start_transition)--> capture_group_start_state -->[inner capture group NFA] --(neg_tagged_transition)--> neg_state --> state_with_positive_tagged_end_transition --(pos_tagged_end_transition)--> end`

SharafMohamed replied:
Added diagram.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexNFAState.hpp:
```diff
@@ -49,9 +49,18 @@ class RegexNFAState {
+    [[nodiscard]] auto get_positive_tagged_end_transitions(
     ) const -> std::vector<PositiveTaggedTransition<RegexNFAState>> const& {
-        return m_positive_tagged_transitions;
+        return m_positive_tagged_end_transitions;
```
Is it possible to have multiple end transitions? From the current implementation it seems like only one single end transition is possible

SharafMohamed replied:
There can only be one end transition. There is a future PR where I make it `optional`, can do it here if you want.

LinZhihao-723 replied:
Yeah I think we should do it here. Having a vector is confusing though

SharafMohamed replied:
Changed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -894,9 +894,16 @@ template <typename NFAStateType>
+    auto* capture_group_start_state = nfa->new_state();
+    root->add_positive_tagged_start_transition(m_tag.get(), capture_group_start_state);
+
     auto* state_with_positive_tagged_transition
```
If the diagram above is correct we should probably rename these states, to be explicit that one is start and one is end

SharafMohamed replied:
Changed.

LinZhihao-723 reviewed (changes requested):
Some comments on the major changes before we proceed

SharafMohamed commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -894,9 +894,16 @@ template <typename NFAStateType>
 void RegexASTCapture<NFAStateType>::add_to_nfa(RegexNFA<NFAStateType>* nfa, NFAStateType* end_state)
         const {
+    NFAStateType* root = nfa->get_root();
+    auto* capture_group_start_state = nfa->new_state();
+    root->add_positive_tagged_start_transition(m_tag.get(), capture_group_start_state);
```
This makes sense for a positive end transition, and negative transition. We changed these in previous PRs to be RAII-style. But, for a positive start transition, it makes sense for a single state to have multiple out-going start transitions. Not sure if there's a good RAII way to do this.

LinZhihao-723 replied:
We could have a factory function to create the start state right? And inside this factory function we should add the new state to the root and mark it as the tagged start transition

SharafMohamed replied:
Ah, I see what you mean based on code rabbit's comment. I didn't realize this is still considered RAII.

LinZhihao-723 replied:
yep yep

SharafMohamed replied:
Changed.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -894,9 +894,15 @@ template <typename NFAStateType>
         const {
+    auto* root = nfa->get_root();
+    auto* capture_group_start_state = nfa->new_state();
+    root->add_positive_tagged_start_transition(m_tag.get(), capture_group_start_state);
```
Changed.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -894,9 +894,21 @@ template <typename NFAStateType>
+            nfa,
+            state_with_positive_tagged_end_transition
+    );
+    nfa->set_root(root);
```
Changed.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -118,6 +131,19 @@ auto RegexNFA<NFAStateType>::new_state_with_negative_tagged_transition(
+    auto* end_state = new_state_with_positive_tagged_transition(tag, dest_state);
+
+    return {start_state, end_state};
+}
```
Fixed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -892,18 +892,19 @@ template <typename NFAStateType>
-    m_group_regex_ast->add_to_nfa_with_negative_tags(nfa, state_with_positive_tagged_transition);
-
-    nfa->set_root(root);
+void RegexASTCapture<NFAStateType>::add_to_nfa(
+        RegexNFA<NFAStateType>* nfa,
+        NFAStateType* dest_state
+) const {
```
```suggestion
auto RegexASTCapture<NFAStateType>::add_to_nfa(
        RegexNFA<NFAStateType>* nfa,
        NFAStateType* dest_state
) const -> void {
```

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexNFAState.hpp:
```diff
@@ -49,9 +49,19 @@ class RegexNFAState {
+        return m_positive_tagged_start_transitions;
+    }
+
+    [[nodiscard]] auto get_positive_tagged_end_transitions(
```
`get_positive_tagged_end_transition`?

SharafMohamed replied:
Yea, singular should be right here.

SharafMohamed replied:
Changed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexNFAState.hpp:
```diff
@@ -176,14 +187,26 @@ auto RegexNFAState<state_type>::serialize(
-        auto const optional_serialized_positive_transition
-                = positive_tagged_transition.serialize(state_ids);
-        if (false == optional_serialized_positive_transition.has_value()) {
+    std::vector<std::string> positive_tagged_start_transition_strings;
```
how about: `serialized_positive_tagged_start_transitions`?

SharafMohamed replied:
Sounds good, this is consistent with our naming convention for other serialized variables.

SharafMohamed replied:
Changed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexNFAState.hpp:
```diff
@@ -176,14 +187,26 @@ auto RegexNFAState<state_type>::serialize(
+        );
+    }
+
+    std::string positive_tagged_end_transition_string;
```
How about: `serialized_positive_tagged_end_transition`?

SharafMohamed replied:
Sounds good, this is consistent with our naming convention for other serialized variables.

SharafMohamed replied:
Changed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -892,11 +892,19 @@ template <typename NFAStateType>
+    // root --(pos_tagged_start_transition)--> capture_group_start_state -->
+    // [inner capture group NFA] --(neg_tagged_transition)--> neg_state -->
+    // state_with_positive_tagged_end_transition --(pos_tagged_end_transition)--> end_state
+    auto [start_state, end_state] = nfa->new_capture_group_start_states(m_tag.get(), dest_state);
```
It would be a bit verbose, but I think we can make it more readable by having a more meaningful name, and sync the name in the comments above:
```suggestion
    auto [positive_tagged_transition_start, positive_tagged_transition_end]
            = nfa->new_positive_tagged_transition_states(m_tag.get(), dest_state);
```

SharafMohamed replied:
Maybe we stick with `capture_group_start_state` and `capture_group_end_state` as implicitly these must have their respective transitions to be the start/end of the capture. And the diagram reinforces this idea for anyone unfamiliar with the NFA of a capture group.

LinZhihao-723 replied:
Sure. But for nfa's factory function, I think we should still call it positive_tagged_transition_states

SharafMohamed replied:
I'm following: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules:
```
Generally speaking, descriptiveness should be proportional to the name's scope of visibility. For example, n may be a fine name within a 5-line function, but within the scope of a class, it's likely too vague.
```
Unless we have a specific guideline that overrules this.
```
template <typename NFAStateType>
auto RegexNFA<NFAStateType>::new_start_and_end_states_with_positively_tagged_transitions(
        Tag const* tag,
        NFAStateType const* dest_state
) -> std::pair<NFAStateType*, NFAStateType*> {
    auto* start_state = new_state();
    m_root->add_positive_tagged_start_transition(tag, start_state);

    auto* end_state = new_state_with_positive_tagged_end_transition(tag, dest_state);
    return {start_state, end_state};
}
```
Within its scope of visibility, there is no ambiguity as to which start and end states these reference.

LinZhihao-723 replied:
1. I'm not suggesting we should change `new_start_and_end_states_with_positively_tagged_transitions`. `start` and `end` states there are fine
2. The reason I commented on `auto [positive_tagged_transition_start, positive_tagged_transition_end]` is for consistency with the diagram in the comment

SharafMohamed replied:
Got it, updated the diagram now that the variable names are fixed.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -118,6 +131,19 @@ auto RegexNFA<NFAStateType>::new_state_with_negative_tagged_transition(
+) -> std::pair<NFAStateType*, NFAStateType*> {
+    auto* start_state = new_state();
+    m_root->add_positive_tagged_start_transition(tag, start_state);
+
+    auto* end_state = new_state_with_positive_tagged_end_transition(tag, dest_state);
+
+    return {start_state, end_state};
```
```suggestion
    auto* end_state = new_state_with_positive_tagged_end_transition(tag, dest_state);
    return {start_state, end_state};
```

Personally I think adding these two empty lines aren't making the code more readable but a little more confusing though...

SharafMohamed replied:
I agree the second line is unneccessary, but I think the first line makes a clear separation b/w the operations pertaining to the `start_state` and the operations pertaining to the `end_state`

LinZhihao-723 replied:
Imo the code is short enough for people to understand stand without empty lines. Adding empty lines in such a short function looks weird

SharafMohamed replied:
I'm going to disagree here. Our general standard so far has been to use newlines to separate logically distinct sections of the code. We create and handle the start state, then we create the end state.

LinZhihao-723 replied:
sure, let's keep the empty line between the creation of two states.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -58,6 +58,19 @@ class RegexNFA {
+     * 2. An end state: end_state --(end `tag`)--> `dest_state`.
+     * @param tag
+     * @param dest_state
+     * @return std::pair<NFAStateType*, NFAStateType*>
```
K, going forward will move more information into the return field.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -58,6 +58,19 @@ class RegexNFA {
+     * @param dest_state
+     * @return std::pair<NFAStateType*, NFAStateType*>
+     */
+    [[nodiscard]] auto new_capture_group_start_states(
```
That name is too general. By encapsulating the capture group behavior (root->start, end->dest) inside the NFA class, we are doing operations very specific to what a capture group would need. If a developer wanted to create new positive tagged transition states, this functionality is already available to them. 

If we don't want capture group specific behavior inside the NFA class, then we're kinda stuck doing what we were doing previously, although I personally think this approach is easier to read and safer.

LinZhihao-723 replied:
Personally I think `new_capture_group_states` sounds like it's creating all the states of the capture group, but it's actually not. I think the name should include some information indicating this is only making the `start` and `end` states

SharafMohamed replied:
I agree.

SharafMohamed replied:
We can kind of combine both our ideas here. Make it correctly describe its functionality, while abstracting out the idea of capture groups: `new_start_and_end_states_with_positively_tagged_transitions`. I think that fully captures the idea of replacing the existing start and end (root and dest) with the new start and end.

SharafMohamed replied:
Changed.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -58,6 +58,19 @@ class RegexNFA {
             NFAStateType const* dest_state
     ) -> NFAStateType*;
 
+    /**
+     * Add two NFA states for a capture group:
+     * 1. A start state: `m_root` --(start `tag`)--> start_state.
+     * 2. An end state: end_state --(end `tag`)--> `dest_state`.
+     * @param tag
+     * @param dest_state
+     * @return std::pair<NFAStateType*, NFAStateType*>
+     */
```
Updated slightly for conciseness:

    /**
     * Creates the start and end states for a capture group.
     * @param tag The tag associated with the capture group.
     * @param dest_state
     * @return A pair of states:
     * - A new state with a positive tagged start transition from `m_root`.
     * - A new state with a positive tagged end transition to `dest_state`.
     */

SharafMohamed replied:
Changed.

SharafMohamed commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -35,13 +35,13 @@ class RegexNFA {
      * `m_states`.
      * @param tag
      * @param dest_state
      * @return NFAStateType*
```
This is out of scope. While I agree the return fields should be better utilized, this was already accepted in a previous PR and not changed in this PR. 

I think the problem with doing these type of small fixes is it misses the bigger picture (e.g., these type of return fields were implemented by me and uncaught in previous PRs). By creating an issue, and then resolving this in a future PR it allows us to do a sweep of the code and fix it everywhere, as opposed to just in this single instance.

LinZhihao-723 replied:
This is not out of scope:
1. It was my mistake to not catch this issue in the previous review
2. The underlying function changed, with both the function name and the underlying thing done by this function. Updating the comment is not out of scope.
3. According to the comments [here](https://github.com/y-scope/log-surgeon/pull/50/files/c757deda4859a787908c1e899213de3fcb423699#r1850917974), we should also update this doc string to be consistent about how the return state is described

SharafMohamed commented on src/log_surgeon/finite_automata/RegexAST.hpp:
```diff
@@ -892,11 +892,19 @@ template <typename NFAStateType>
+        RegexNFA<NFAStateType>* nfa,
+        NFAStateType* dest_state
+) const {
+    // root --(pos_tagged_start_transition)--> capture_group_start_state -->
+    // [inner capture group NFA] --(neg_tagged_transition)--> neg_state -->
+    // state_with_positive_tagged_end_transition --(pos_tagged_end_transition)--> end_state
```
Will update this once we settle on variable naming.

SharafMohamed replied:
Updated.

LinZhihao-723 reviewed (commented):
I think the major renaming issue is about this new graph we introduced:
```
root --(pos_tagged_start_transition)--> capture_group_start_state --> [inner capture group NFA] --(neg_tagged_transition)--> neg_state --> capture_group_end_state --(pos_tagged_end_transition)--> end_state
```
Things we introduced:
- pos_tagged_start_transition (edge)
- pos_tagged_end_transition (edge)
- capture_group_start_state (state)
- capture_group_end_state (state)

IMO the naming of these four edges/states is important for people who first read this code. I'm not sure if the current naming is already optimized to its best state:
- positive "tagged" start/end transition and "capture group" states using different terms. Is it possible to stick to one? Either `tagged` or `capture group`
- If we use the transition to name a new state, i.e., `new_state_with_positive_tagged_end_transition` isn't clear which state it creates, as it has two states connected. And for people who first read the code, it could be either the transition points to the new state or the new state has a new transition pointing to other state. Is it more easier to understand if we explicitly name the `new` function by its state, i.e., `new_capture_group_end_state`, that matches the diagram above and can visually tell people which state it creates. (that is why I suggested `new_positive_tagged_transition_end_state` as it explicitly speaks out which new state it creates, but we may need to come up a better name rather than `positive_tagged_transition_end_state`).

It's my mistake that I should realize some of the problems above in the previous PR. When this PR introduces more complexity to the tagged transitions, I realized more clarifications might be required.
I think the action item would be to formalize the diagram above, and then we can apply to the related code and future PRs. Maybe we can send an issue to track and get this PR merged first.
@kirkrodrigues Can u help take a look as well?

SharafMohamed commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -35,13 +35,13 @@ class RegexNFA {
      * @return NFAStateType*
      */
-    [[nodiscard]] auto new_state_with_positive_tagged_transition(
+    [[nodiscard]] auto new_state_with_positive_tagged_end_transition(
```
Sticking with the `new_state_with_X` convention used previously.

LinZhihao-723 replied:
`new_state_with_positively_tagged_end_transition` to be consistent with `new_start_and_end_states_with_positively_tagged_transitions`?

SharafMohamed replied:
Kk issue created.

SharafMohamed commented:
> I think the major renaming issue is about this new graph we introduced:
> 
> ```
> root --(pos_tagged_start_transition)--> capture_group_start_state --> [inner capture group NFA] --(neg_tagged_transition)--> neg_state --> capture_group_end_state --(pos_tagged_end_transition)--> end_state
> ```
> 
> Things we introduced:
> 
> * pos_tagged_start_transition (edge)
> * pos_tagged_end_transition (edge)
> * capture_group_start_state (state)
> * capture_group_end_state (state)
> 
> IMO the naming of these four edges/states is important for people who first read this code. I'm not sure if the current naming is already optimized to its best state:
> 
> * positive "tagged" start/end transition and "capture group" states using different terms. Is it possible to stick to one? Either `tagged` or `capture group`
> * If we use the transition to name a new state, i.e., `new_state_with_positive_tagged_end_transition` isn't clear which state it creates, as it has two states connected. And for people who first read the code, it could be either the transition points to the new state or the new state has a new transition pointing to other state. Is it more easier to understand if we explicitly name the `new` function by its state, i.e., `new_capture_group_end_state`, that matches the diagram above and can visually tell people which state it creates. (that is why I suggested `new_positive_tagged_transition_end_state` as it explicitly speaks out which new state it creates, but we may need to come up a better name rather than `positive_tagged_transition_end_state`).
> 
> It's my mistake that I should realize some of the problems above in the previous PR. When this PR introduces more complexity to the tagged transitions, I realized more clarifications might be required. I think the action item would be to formalize the diagram above, and then we can apply to the related code and future PRs. Maybe we can send an issue to track and get this PR merged first. @kirkrodrigues Can u help take a look as well?

```
root --(`m_tag` start)--> capture_start_state --> [`m_group_regex_ast` NFA] --(`m_negative_tags`)--> capture_end_state --(`m_tag` end)--> dest_state
```
This diagram uses concise variable names, and directly uses variable name for all transitions, states, and the sub-NFA. In terms of incoming vs outgoing edges, in general LS has states store out-going edges, therefore any edge associated with a state is always out-going. Maybe this can be made more clear in a a docstring for the NFAState class.

LinZhihao-723 commented on src/log_surgeon/finite_automata/RegexNFA.hpp:
```diff
@@ -58,6 +58,19 @@ class RegexNFA {
+     * - A new state with a positive tagged start transition from `m_root`.
+     * - A new state with a positive tagged end transition to `dest_state`.
+     */
+    [[nodiscard]] auto new_start_and_end_states_with_positively_tagged_transitions(
```
If we use `positively` instead of `positive,` we should propagate this change to all the occurrences. Not sure if we should do it in this PR, maybe send a issue to change it in the future PR, and keep this function `new_start_and_end_states_with_positive_tagged_transitions`

SharafMohamed replied:
This should be `new_start_and_end_states_with_positive_tagged_transitions`, positively was a typo.

SharafMohamed replied:
Changed.

LinZhihao-723 reviewed (approved):
for the PR title, how about:
`feat: Split NFA positive tags into start and end transitions to encapsulate a capture group.`
